File: source/ValidPalindrome.cpp
#include <string>

class Solution
{
public:
    bool isPalindrome(std::string s)
    {
        if (s.empty() || s.size() == 1) return true;

        std::string::iterator front = s.begin();
        std::string::iterator back = s.end();

        while (front < back)
        {
            while (front != s.end() && !std::isalnum(*front)) ++front;
            while (back != s.begin() && !std::isalnum(*back)) --back;

            if (front < back && std::tolower(*front) != std::tolower(*back))
            {
                return false;
            }

            ++front;
            --back;
        }

        return true;
    }
};

File: CSharpPractice/source/ValidPalindrome.cs
// https://leetcode.com/problems/valid-palindrome

namespace LeetCodePractice.ValidPalindrome
{
    public class Solution
    {
        public bool IsPalindrome(string s)
        {
            if (s.Length == 0 || s.Length == 1) return true;

            int front = 0;
            int back = s.Length - 1;

            while (front < back)
            {
                while (front < s.Length && !char.IsLetterOrDigit(s[front])) ++front;
                while (back >= 0 && !char.IsLetterOrDigit(s[back])) --back;

                if (front < back && char.ToLower(s[front]) != char.ToLower(s[back]))
                {
                    return false;
                }

                ++front;
                --back;
            }

            return true;
        }
    }
}